### PR TITLE
datetime - validate unpacked data

### LINKED ISSUE #6273: net.replicaset
A connector to one replicaset. It has the following abilities:

- Master-discovering functional inside
- Traffic weight/zone-oriented balancer

```lua
box.cfg{zone='moscow', weight=45}   -- current zone and default weight

replicaset = require 'net.replicaset'

rs = replicaset.connect{
    url = {
        'user:password@host1',
        'user:password@host2',

    },

    pri = {
          moscow = { -- weights
               spb = 5,
               moscow = 10000000,
               vladivostok = nil,      -- access only if the other hosts are down
          }
    }
    -- какие-то опции
    reconect_after = 0.1,
}

-- API

rs.space.name:get_anon{...} -- anonimous replica is preferred
rs.space.name:get_rw{...} -- master
rs.space.name:get{...} -- ro replica is preferred

rs:call_anon(...)
rs:call_rw(...)
rs:call(...)

-- master
rs.space.name:replace{}
rs.space.name:insert{...}
rs.space.name:update{...}

rs:close()
```

### PR DESCRIPTION
**Pull request by tsafin**
Introduced few checks of validity of incoming data
at the moment messagepack is converted to datetime.

- we check that size is either big enough for single
  epoch or whole record;
- and we check range of data in decoded fields.

Fixes #6273

### CODE REVIEW

**Comment by tsafin** [comment]
Discovered the harder way that `checkpatch` is not yet ignoring binary files to be added to fuzzer corpus. Opened https://github.com/tarantool/checkpatch/issues/21 for that.

File: src/lib/core/datetime.h [hunk 1]
```diff
@@ -9,6 +9,7 @@
 #include <stdint.h>
```
Comment (sergepetrenko) [inline src/lib/core/datetime.h:9]: commit "datetime: validate unpacked data"

> NO_DOC=bugfix
NO_CHANGELOG=bugfix
NO_TEST=see the other commit for fuzzer

I'd rather add a changelog.

File: src/lib/core/datetime.h [hunk 2]
```diff
@@ -400,6 +401,18 @@ datetime_usec(const struct datetime *date)
 	return datetime_nsec(date) / 1000;
 }
 
+/** Simplified checker for datetime structure validity */
+static inline bool
+datetime_validate(const struct datetime *date)
+{
+	if (unlikely(date->epoch < MIN_EPOCH_SECS_VALUE) ||
```
Comment (sergepetrenko) [inline src/lib/core/datetime.h:409]: Is there any chance we can validate `tzindex`?

**Review by sergepetrenko — CHANGES_REQUESTED** [review]
Hi! Thanks for the patch! Please, find my comments below.

File: src/lib/core/datetime.h (hunk 2 above)
Comment (tsafin) [reply src/lib/core/datetime.h:409]: Yup, there is known maximum for tzcode index we generated. But it's in the otehr compilation unit, so we need to invent some trick to reuse without circle dependency.

File: src/lib/core/datetime.h [hunk 3]
```diff
@@ -400,6 +401,18 @@ datetime_usec(const struct datetime *date)
...
+	    unlikely(date->epoch > MAX_EPOCH_SECS_VALUE))
+		return false;
```
[the remark this answers is not in the document]
Comment (tsafin) [reply src/lib/core/datetime.h:411]: Good suggestion!

[the remark this answers is not in the document]
Comment (tsafin) [reply src/lib/core/datetime.h:411]:
```diff
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index 725b500ff..63fe45962 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -62,6 +62,12 @@ struct tnt_tm;
 #define MIN_EPOCH_SECS_VALUE    \
 	(MIN_DT_DAY_VALUE * SECS_PER_DAY - SECS_EPOCH_1970_OFFSET)
 
+/** at the moment the range of known timezones is UTC-12:00..UTC+14:00
+ * https://en.wikipedia.org/wiki/List_of_UTC_time_offsets
+  */
+#define MAX_TZOFFSET (14L*60)
+#define MIN_TZOFFSET (-12L*60)
+
 /**
  * datetime structure keeps number of seconds and
  * nanoseconds since Unix Epoch.
@@ -411,6 +417,9 @@ datetime_validate(const struct datetime *date)
 		return false;
 	if (unlikely(date->nsec < 0) || unlikely(date->nsec > NANOS_PER_SEC))
 		return false;
+	if (unlikely(date->tzoffset < MIN_TZOFFSET) ||
+	    unlikely(date->tzoffset > MAX_TZOFFSET))
+		return false;
 	return true;
 }
 
```

File: src/lib/core/datetime.h (hunk 2 above)
Comment (tsafin) [reply src/lib/core/datetime.h:409]:
```diff
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index 63fe45962..88ea95732 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -62,12 +62,19 @@ struct tnt_tm;
 #define MAX_TZOFFSET (14L*60)
 #define MIN_TZOFFSET (-12L*60)
 
+/**
+ * Actually we have lesser number of generated timezones, but 1024
+ * might be a good estimate.
+ */
+#define MAX_TZINDEX 1024
+
 /**
  * datetime structure keeps number of seconds and
  * nanoseconds since Unix Epoch.
@@ -420,6 +427,9 @@ datetime_validate(const struct datetime *date)
 	if (unlikely(date->tzoffset < MIN_TZOFFSET) ||
 	    unlikely(date->tzoffset > MAX_TZOFFSET))
 		return false;
+	if (unlikely(date->tzindex < 0) ||
+	    unlikely(date->tzindex > MAX_TZINDEX))
+		return false;
 	return true;
 }
 
diff --git a/src/lib/tzcode/timezone.c b/src/lib/tzcode/timezone.c
index 4f1468291..d86f4f634 100644
--- a/src/lib/tzcode/timezone.c
+++ b/src/lib/tzcode/timezone.c
@@ -25,15 +25,14 @@ static struct date_time_zone zones_raw[] = {
 #undef ZONE_ABBREV
 };
 
-#define ZONES_MAX 1024
-static_assert(ZONES_MAX >= lengthof(zones_raw), "please increase ZONES_MAX");
+static_assert(MAX_TZINDEX >= lengthof(zones_raw), "please increase MAX_TZINDEX");
 
 /**
  * Sorted array of zone descriptors, whether it's abbreviation, full
  * zone name or [backward-compatible] link name.
  */
 static struct date_time_zone zones_sorted[lengthof(zones_raw)];
-static struct date_time_zone zones_unsorted[ZONES_MAX];
+static struct date_time_zone zones_unsorted[MAX_TZINDEX];
 
 static int
 compare_zones(const void *a, const void *b)

```

File: src/lib/core/datetime.h [hunk 4]
```diff
@@ -400,6 +415,24 @@ datetime_usec(const struct datetime *date)
...
+	if (unlikely(date->nsec < 0) || unlikely(date->nsec > NANOS_PER_SEC))
```
Comment (sergepetrenko) [inline src/lib/core/datetime.h:425]: IIRC, there were some cases (maybe not present now, but planned), when date->nsec could be `2 000 000 000` due to a leap second.
Should we allow it here? Or did I mix things up?

**Review by sergepetrenko — COMMENTED** [review]
Hi! Thanks for the fixes!
All good except one comment bothering me.

File: src/lib/core/datetime.h (hunk 4 above)
Comment (tsafin) [reply src/lib/core/datetime.h:425]: That's very good question: yes, originally we assumed that once we would need to handle some specific leap second then it will be done via nanoseconds part which would be allowed to handle 2e9 nanoseconds.

But during integration of tzcode in previous stages we have discovered that if we would allow leap-seconds in tzcode epoch then we would receive epoch values different to system. (And in POSIX epoch we do not save leap-seconds at all, they are non-existing in the range of posix values). tzcode, if being compiled with precise leap-seconds support, does have routine to convert between their own epoch values and system epoch, with skipped leap-seconds. But, for simplicity sake, we have opted to not behave differently to the surrounding system, and behave as it's done in posix functions.

That's why there is no currently way to receive automatically constructed/parsed date/time value where leap-seconds would be larger than 1e9. (And if that's incorrect we do have a couple of assertions here and there). That's why it may be still ok to accept only upto 1e9, not 2e9.

But in a longer term yes, we need to create `MAX_NANOS_PER_SEC` as 2e9, and apply it here and in parsers.

Comment (tsafin) [reply src/lib/core/datetime.h:425]: Not sure (yet) it's a blocker for current pulle-request.

Comment (sergepetrenko) [reply src/lib/core/datetime.h:425]: Ok. What happens if we allow nsec up to `2e9` now and receive a date with `nsec == 2e9`? Will some assertion fire once we try to process it?

If everything will be fine, I suggest to allow `2e9` in nsec right now. So that possible future extension is accepted by current Tarantool versions.

If something will break right now when we try to process `2e9`, then let's disallow it.

Comment (tsafin) [reply src/lib/core/datetime.h:425]:
```diff
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index 8d8d96da9..6f41c2f3a 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -32,6 +32,7 @@ struct tnt_tm;
 #define SECS_PER_DAY          86400
 #define DT_EPOCH_1970_OFFSET  719163
 #define NANOS_PER_SEC         1000000000LL
+#define MAX_NANOS_PER_SEC     2000000000LL
 #endif
 
 /** Required size of datetime_to_string string buffer */
@@ -422,7 +423,8 @@ datetime_validate(const struct datetime *date)
 	if (unlikely(date->epoch < MIN_EPOCH_SECS_VALUE) ||
 	    unlikely(date->epoch > MAX_EPOCH_SECS_VALUE))
 		return false;
-	if (unlikely(date->nsec < 0) || unlikely(date->nsec > NANOS_PER_SEC))
+	if (unlikely(date->nsec < 0) ||
+	    unlikely(date->nsec > MAX_NANOS_PER_SEC))
 		return false;
 	if (unlikely(date->tzoffset < MIN_TZOFFSET) ||
 	    unlikely(date->tzoffset > MAX_TZOFFSET))
diff --git a/src/lib/tzcode/strftime.c b/src/lib/tzcode/strftime.c
index b6bfbf1ea..9bb437e24 100644
--- a/src/lib/tzcode/strftime.c
+++ b/src/lib/tzcode/strftime.c
@@ -35,6 +35,7 @@
 
 #include "private.h"
 #include "trivia/util.h"
+#include "datetime.h"
 #include "tzcode.h"
 #include "timezone.h"
 #include "timelocal.h"
@@ -281,7 +282,7 @@ _fmt(char *buf, ssize_t size, const char *format, const struct tnt_tm *t,
 				/* fallthru */
 			case 'f': {
 				int nsec = t->tm_nsec;
-				assert(nsec < 1000000000);
+				assert(nsec < MAX_NANOS_PER_SEC);
 
 				/*
 				 ** no length modifier provided -
```

**Review by sergepetrenko — APPROVED** [review]
Thanks for working on this! LGTM.

File: test/static/corpus/mp_datetime/889c6a675df321355334e553ff014b259ba63cf3 [hunk 5]
```diff
@@ -0,0 +1 @@
+jj�������o����
```
[the remark this answers is not in the document]
Comment (tsafin) [reply test/static/corpus/mp_datetime/889c6a675df321355334e553ff014b259ba63cf3:1]: Шайтан!
image

[the remark this answers is not in the document]
Comment (tsafin) [reply test/static/corpus/mp_datetime/889c6a675df321355334e553ff014b259ba63cf3:1]:
```diff
diff --git a/test/static/corpus/mp_datetime/1ceff235fd5b31eaa9bf88ca610ca924c2b359ad b/test/static/corpus/mp_datetime/1ceff235fd5b31eaa9bf88ca610ca924c2b359ad
deleted file mode 100644
index 0fc102596..000000000
Binary files a/test/static/corpus/mp_datetime/1ceff235fd5b31eaa9bf88ca610ca924c2b359ad and /dev/null differ
diff --git a/test/static/corpus/mp_datetime/5a4d479cd108238221d5f2026002fbb9719f0a1a b/test/static/corpus/mp_datetime/5a4d479cd108238221d5f2026002fbb9719f0a1a
deleted file mode 100644
index 436b922b0..000000000
Binary files a/test/static/corpus/mp_datetime/5a4d479cd108238221d5f2026002fbb9719f0a1a and /dev/null differ

```

[the remark this answers is not in the document]
Comment (tsafin) [reply test/static/corpus/mp_datetime/889c6a675df321355334e553ff014b259ba63cf3:1]: Also, though, as a bonus to previos `MAX_NANOS_PER_SEC` changes I introduce this change to the SQL-related code, to make it behave consistently with outer `datetime.c` code:

```diff
diff --git a/src/lib/core/datetime.c b/src/lib/core/datetime.c
index 5b6587c90..a2f206482 100644
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -1045,7 +1045,7 @@ datetime_from_fields(struct datetime *dt, const struct dt_fields *fields)
 		return -1;
 	double nsec = fields->msec * 1000000 + fields->usec * 1000 +
 		      fields->nsec;
-	if (nsec < 0 || nsec > NANOS_PER_SEC)
+	if (nsec < 0 || nsec > MAX_NANOS_PER_SEC)
 		return -1;
 	if (fields->tzoffset < -720 || fields->tzoffset > 840)
 		return -1;
```

[the remark this answers is not in the document]
Comment (tsafin) [reply test/static/corpus/mp_datetime/889c6a675df321355334e553ff014b259ba63cf3:1]: After this change introduced above I've received failures in `sql-luatest/datetime_test.lua`, so we need to revisit it once again, and change test (and original code as well, as test showed incorrect boundary checks - we allow range `[0 .. 2e9)` but not `[0 .. 2e9]`)

```diff
diff --git a/src/lib/core/datetime.c b/src/lib/core/datetime.c
index a2f206482..ab4f5ce6e 100644
--- a/src/lib/core/datetime.c
+++ b/src/lib/core/datetime.c
@@ -1045,7 +1045,7 @@ datetime_from_fields(struct datetime *dt, const struct dt_fields *fields)
 		return -1;
 	double nsec = fields->msec * 1000000 + fields->usec * 1000 +
 		      fields->nsec;
-	if (nsec < 0 || nsec > MAX_NANOS_PER_SEC)
+	if (nsec < 0 || nsec >= MAX_NANOS_PER_SEC)
 		return -1;
 	if (fields->tzoffset < -720 || fields->tzoffset > 840)
 		return -1;
diff --git a/src/lib/core/datetime.h b/src/lib/core/datetime.h
index 6f41c2f3a..fcf1c7aea 100644
--- a/src/lib/core/datetime.h
+++ b/src/lib/core/datetime.h
@@ -424,7 +424,7 @@ datetime_validate(const struct datetime *date)
 	    unlikely(date->epoch > MAX_EPOCH_SECS_VALUE))
 		return false;
 	if (unlikely(date->nsec < 0) ||
-	    unlikely(date->nsec > MAX_NANOS_PER_SEC))
+	    unlikely(date->nsec >= MAX_NANOS_PER_SEC))
 		return false;
 	if (unlikely(date->tzoffset < MIN_TZOFFSET) ||
 	    unlikely(date->tzoffset > MAX_TZOFFSET))
diff --git a/test/sql-luatest/datetime_test.lua b/test/sql-luatest/datetime_test.lua
index 70bdd6ad4..64b4e64ff 100644
--- a/test/sql-luatest/datetime_test.lua
+++ b/test/sql-luatest/datetime_test.lua
@@ -2527,9 +2527,9 @@ g.test_datetime_33_2 = function()
         t.assert_equals(err.message, res)
 
         -- "msec" cannot be more than 1000.
-        v = {msec = 1001}
+        v = {msec = 2000}
         _, err = box.execute(sql, {{['#v'] = v}})
-        res = [[Type mismatch: can not convert map({"msec": 1001}) to datetime]]
+        res = [[Type mismatch: can not convert map({"msec": 2000}) to datetime]]
         t.assert_equals(err.message, res)
 
         -- "msec" cannot be less than 0.
@@ -2539,9 +2539,9 @@ g.test_datetime_33_2 = function()
         t.assert_equals(err.message, res)
 
         -- "usec" cannot be more than 1000000.
-        v = {usec = 1000001}
+        v = {usec = 2000000}
         _, err = box.execute(sql, {{['#v'] = v}})
-        res = [[Type mismatch: can not convert map({"usec": 1000001}) ]]..
+        res = [[Type mismatch: can not convert map({"usec": 2000000}) ]]..
               [[to datetime]]
         t.assert_equals(err.message, res)
 
@@ -2552,9 +2552,9 @@ g.test_datetime_33_2 = function()
         t.assert_equals(err.message, res)
 
         -- "nsec" cannot be more than 1000000000.
-        v = {nsec = 1000000001}
+        v = {nsec = 2000000000}
         _, err = box.execute(sql, {{['#v'] = v}})
-        res = [[Type mismatch: can not convert map({"nsec": 1000000001}) ]]..
+        res = [[Type mismatch: can not convert map({"nsec": 2000000000}) ]]..
               [[to datetime]]
         t.assert_equals(err.message, res)
```

File: test/fuzz/CMakeLists.txt [hunk 6]
```diff
@@ -50,9 +50,13 @@ target_link_libraries(datetime_parse_full_fuzzer PUBLIC core fuzzer_config)
 add_executable(datetime_strptime_fuzzer datetime_strptime_fuzzer.c)
 target_link_libraries(datetime_strptime_fuzzer PUBLIC core fuzzer_config)
 
+add_executable(mp_datetime_fuzzer mp_datetime_fuzzer.c)
+target_link_libraries(mp_datetime_fuzzer PUBLIC core fuzzer_config)
+
 set(fuzzing_binaries csv_fuzzer
                      datetime_parse_full_fuzzer
                      datetime_strptime_fuzzer
+                     mp_datetime_fuzzer
```
[the remark this answers is not in the document]
Comment (tsafin) [reply test/fuzz/CMakeLists.txt]: Done

```diff
--- a/test/fuzz/CMakeLists.txt
+++ b/test/fuzz/CMakeLists.txt
@@ -56,8 +56,8 @@ target_link_libraries(mp_datetime_fuzzer PUBLIC core fuzzer_config)
 set(fuzzing_binaries csv_fuzzer
                      datetime_parse_full_fuzzer
                      datetime_strptime_fuzzer
-                     mp_datetime_fuzzer
                      http_parser_fuzzer
+                     mp_datetime_fuzzer
                      swim_proto_member_fuzzer
                      swim_proto_meta_fuzzer
                      uri_fuzzer)
```

File: src/lib/core/mp_datetime.c [hunk 7]
```diff
@@ -55,7 +55,7 @@ mp_sizeof_datetime(const struct datetime *date)
 struct datetime *
 datetime_unpack(const char **data, uint32_t len, struct datetime *date)
```
[the remark this answers is not in the document]
Comment (tsafin) [reply src/lib/core/mp_datetime.c:56]: I'me very confused with that your request, as I don't know how to control access to boundary conditions without my own mutator (which I don't want to write yet here). My understanding is that it's fuzzer infrastructure responsibility to mutate input `data` stream, and eventually (given infinite time) it would hit boundary condition values. 

More efficiently  boundary values are being checked in `app-tap/datetime.test.lua`, for all fields accessible from constructors. And it look good enough for today.

[the remark this answers is not in the document]
Comment (tsafin) [reply src/lib/core/mp_datetime.c:56]: Created something as simple as that (in the datetime unit test code):
```diff
diff --git a/test/unit/datetime.c b/test/unit/datetime.c
index 1dd7c59ac..4578ae8fc 100644
--- a/test/unit/datetime.c
+++ b/test/unit/datetime.c
@@ -440,6 +440,66 @@ mp_datetime_test()
 	check_plan();
 }
 
+static void
+mp_datetime_unpack_valid_checks(void)
+{
+	/* Binary, message-pack representation of datetime
+	 * payload in MP value */
+	struct binary_datetime {
+		/** Seconds since Epoch. */
+		int64_t epoch;
+		/** Nanoseconds, if any. */
+		int32_t nsec;
+		/** Offset in minutes from UTC. */
+		int16_t tzoffset;
+		/** Olson timezone id */
+		int16_t tzindex;
+	};
+
+	static struct binary_datetime invalid_values[] = {
+		{.epoch = MAX_EPOCH_SECS_VALUE + 1},
+		{.epoch = MIN_EPOCH_SECS_VALUE - 1},
+		{.nsec = MAX_NANOS_PER_SEC},
+		{.nsec = -1},
+		{.tzoffset = MIN_TZOFFSET - 1},
+		{.tzoffset = MAX_TZOFFSET + 1},
+		{.tzindex = MAX_TZINDEX + 1},
+		{.tzindex = -1},
+	};
+
+	static struct binary_datetime valid_values[] = {
+		{.epoch = MAX_EPOCH_SECS_VALUE},
+		{.epoch = MIN_EPOCH_SECS_VALUE},
+		{.nsec = MAX_NANOS_PER_SEC - 1},
+		{.nsec = 0},
+		{.tzoffset = MIN_TZOFFSET},
+		{.tzoffset = MAX_TZOFFSET},
+		{.tzindex = MAX_TZINDEX},
+		{.tzindex = 0},
+	};
+	size_t index;
+	const char* p;
+	struct datetime date;
+
+	plan(24);
+	for (index = 0; index < lengthof(valid_values); index++) {
+		struct binary_datetime value = valid_values[index];
+		p = (char*)&value;
+		memset(&date, 0, sizeof(date));
+		struct datetime *dt = datetime_unpack(&p, sizeof(value), &date);
+		isnt(dt, NULL, "datetime_unpack() is not NULL");
+		is((int64_t)dt->epoch, value.epoch, "epoch value expected");
+	}
+
+	for (index = 0; index < lengthof(valid_values); index++) {
+		struct binary_datetime value = invalid_values[index];
+		p = (char*)&value;
+		memset(&date, 0, sizeof(date));
+		struct datetime *dt = datetime_unpack(&p, sizeof(value), &date);
+		is(dt, NULL, "datetime_unpack() is NULL");
+	}
+	check_plan();
+}
 
 static int
 mp_fprint_ext_test(FILE *file, const char **data, int depth)
@@ -505,10 +565,11 @@ mp_print_test(void)
 int
 main(void)
 {
-	plan(5);
+	plan(6);
 	datetime_test();
 	tostring_datetime_test();
 	parse_date_test();
+	mp_datetime_unpack_valid_checks();
 	mp_datetime_test();
 	mp_print_test();
```

**Comment by kyukhin** [comment]
Backported to 2.10 as well.
